Source/JavaScriptCore/runtime/JSObjectInlines.h

```diff
@@ -904,31 +904,50 @@ void JSObject::forEachOwnIndexedProperty(JSGlobalObject* globalObject, const Fun
     }
 
     case ALL_ARRAY_STORAGE_INDEXING_TYPES: {
-        Vector<unsigned, 8> properties;
-        MarkedArgumentBuffer values;
-
         ArrayStorage* storage = m_butterfly->arrayStorage();
         unsigned usedVectorLength = std::min(storage->length(), storage->vectorLength());
         for (unsigned i = 0; i < usedVectorLength; ++i) {
             auto value = storage->m_vector[i];
-            if (value) {
-                properties.append(i);
-                values.appendWithCrashOnOverflow(value.get());
-            }
+            if (!value)
+                continue;
+            if (functor(i, value.get()) == IterationStatus::Done)
+                return;
         }
 
         if (SparseArrayValueMap* map = storage->m_sparseMap.get()) {
-            for (auto& [key, value] : *map) {
-                if (!(value.attributes() & PropertyAttribute::DontEnum)) {
-                    properties.append(key);
-                    values.appendWithCrashOnOverflow(value.get());
+            Vector<unsigned, 8> properties;
+            MarkedArgumentBuffer values;
+            if constexpr (mode == JSObject::SortMode::Default) {
+                for (auto& [key, value] : *map) {
+                    if (!(value.attributes() & PropertyAttribute::DontEnum)) {
+                        properties.append(key);
+                        values.appendWithCrashOnOverflow(value.get());
+                    }
                 }
-            }
-        }
 
-        for (size_t i = 0; i < properties.size(); ++i) {
-            if (functor(properties[i], values.at(i)) == IterationStatus::Done)
-                return;
+                for (size_t i = 0; i < properties.size(); ++i) {
+                    if (functor(properties[i], values.at(i)) == IterationStatus::Done)
+                        return;
+                }
+            } else {
+                HashMap<unsigned, unsigned, WTF::IntHash<unsigned>, WTF::UnsignedWithZeroKeyHashTraits<unsigned>> propertyToValueIndex;
+                unsigned valueIndex = 0;
+                for (auto& [key, value] : *map) {
+                    if (!(value.attributes() & PropertyAttribute::DontEnum)) {
+                        properties.append(key);
+                        values.appendWithCrashOnOverflow(value.get());
+                        propertyToValueIndex.add(key, valueIndex++);
+                    }
+                }
+
+                std::sort(properties.begin(), properties.end());
+                for (size_t i = 0; i < properties.size(); ++i) {
+                    unsigned property = properties.at(i);
+                    unsigned valueIndex = propertyToValueIndex.get(property);
+                    if (functor(property, values.at(valueIndex)) == IterationStatus::Done)
+                        return;
+                }
+            }
```

Comment:
This looks super costly. Why not

1. Creating properties with `std::tuple<unsigned, unsigned>`, index and valueIndex
2. Sort it with the first key
3. `auto [index, valueIndex] = ...; functor(index, values.at(valueIndex))`